### PR TITLE
Adds live plot context manager and update events

### PR DESCRIPTION
**Pull request by StephenNneji**
Adds a live plot as a context manager, to use change the controls 
```
controls.procedure = 'simplex';
controls.maxIterations = 100;
controls.updatePlotFreq = 1;
```
then run like so 
```
with RAT.plotting.LivePlot(block=True):
    problem, results = RAT.run(problem, controls)
```
or 

```
with RAT.plotting.LivePlot(block=True) as figure:
    problem, results = RAT.run(problem, controls)
```

### CODE REVIEW

**Review by DrPaulSharp — COMMENTED** [review]
I can confirm that the plots behave as expected and discussed both when importing and running the examples as a script. I have tested this using both pycharm and the windows x64 command prompt.

Alongside the comments could you please review the test coverage for "plotting.py", in particular could you please write a test for "plot_ref_sld".

File: tests/test_plotting.py [hunk 1]
```diff
@@ -31,21 +31,20 @@ def data() -> PlotEventData:
...
 
 @pytest.fixture
-def fig() -> Figure:
+def fig() -> plt.figure:
     """Creates the fixture for the tests."""
     plt.close("all")
-    figure = Figure(1, 3)
-    fig = plot_ref_sld_helper(fig=figure, data=data())
-    return fig
+    figure = plt.subplots(1, 3)[0]
```
[the remark this answers is not in the document]
Comment (StephenNneji) [reply tests/test_plotting.py:37]: Yes, its a check that the plots are corrected

File: RATapi/utils/plotting.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 from typing import Optional, Union
 
+import matplotlib
```
[the remark this answers is not in the document]
Comment (StephenNneji) [reply RATapi/utils/plotting.py:5]: No, removed

File: RATapi/utils/plotting.py [hunk 3]
```diff
@@ -75,33 +37,38 @@ def plot_errorbars(ax: Axes, x: np.ndarray, y: np.ndarray, err: np.ndarray, one_
     ax.scatter(x=x, y=y, s=3, marker="o", color=color)
 
 
-def plot_ref_sld_helper(data: PlotEventData, fig: Optional[Figure] = None, delay: bool = True):
+def plot_ref_sld_helper(data: PlotEventData, fig: Optional[matplotlib.pyplot.figure] = None, delay: bool = True):
```
[the remark this answers is not in the document]
Comment (StephenNneji) [reply RATapi/utils/plotting.py:40]: Is this the intended outcome of the labels or am I missing something? Also are we sure the units will always be Angstrom?
Screenshot 2024-07-19 095339

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RATapi/utils/plotting.py:40]: Ahh, so close. The ^ and _ are meant to be superscripts and subscripts.  good point about units we should check with @arwelHughes, but I'm not aware of how one would change the units in the code.

[the remark this answers is not in the document]
Comment (StephenNneji) [reply RATapi/utils/plotting.py:40]: Also updating the labels via PlotEventData will require me diving into the RAT, C++ to add more stuff so I will do that in another PR but I will update the labels and add a unit test for plot_ref_sld

[the remark this answers is not in the document]
Comment (StephenNneji) [reply RATapi/utils/plotting.py:40]: > Ahh, so close. The ^ and _ are meant to be superscripts and subscripts. 

This works
ref_plot.set_xlabel("$q_{z} (\u00c5^{-1})$")
ref_plot.set_ylabel("$Ref (q_{z})$")
sld_plot.set_xlabel("$Z (\u00c5)$")
sld_plot.set_ylabel("$SLD (\u00c5^{-2})$")

[the remark this answers is not in the document]
Comment (StephenNneji) [reply RATapi/utils/plotting.py:40]: Screenshot 2024-07-19 100928

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RATapi/utils/plotting.py:40]: Ahh, no I had it that the reflectivity is a function of qz, but obviously it's a bit confusing like this. Maybe best to leave that off then.

File: tests/test_plotting.py [hunk 4]
```diff
@@ -137,3 +137,26 @@ def test_live_plot(mock: MagicMock) -> None:
     assert mock.call_args_list[2].args[2] == 0.0
     assert mock.call_args_list[2].args[4] == 153
     assert mock.call_args_list[2].args[5] == 1.0
+
+
+@patch("RATapi.utils.plotting.plot_ref_sld_helper")
+def test_plot_ref_sld(mock: MagicMock, request) -> None:
+    problem = request.getfixturevalue("input_project")
+    results = request.getfixturevalue("reflectivity_calculation_results")
```
Comment (DrPaulSharp) [inline tests/test_plotting.py:145]: The strongly recommended way to request fixtures is to include the them in the function arguments. This method of requesting fixtures is for when the choice of fixture depends on other parameters. I've seen that line 554 of "test_run.py" has a line like this as well (no idea who wrote it . . .) - could you please also change that to include the fixture in the function argument.

**Review by DrPaulSharp — APPROVED** [review]

**Pull request merged**